scanner/decisions/0005-scannerv4-read-cvss-data.md

```diff
@@ -0,0 +1,27 @@
+# 0005 - Adapt Scanner V4 NVD CVSS data updating pipeline to integrate with NVD CVE Api
+
+- **Author(s):** Yi Li <yli3@redhat.com>
+- **Created:** [2023-10-10]
+
+## Status
+
+Accepted
+
+## Context
+
+The National Vulnerability Database (NVD) has announced the retirement of its JSON feeds by December 15, 2023. This creates the need to change our existing CVSS data updater pipeline for Scanner V4, which currently relies on these JSON feeds.
+
+## Decision
+
+The current NVD CVSS GitHub Workflow will be updated to fetch data from the NVD CVE API instead of NVD Json feeds while maintaining the freshness of the NVD CVSS data bundle in Google Storage. It will categorize the downloaded data by CVSS V3 severity levels, and these categorized data will be archived into individual zip files, such as `severity-low.zip`. The format of the JSON is the same as before.
+
+The CVSS Updater in Central will be modified to download zip files corresponding to four distinct severity levels and write them into a single zip file stored in Central's file system. This eliminates the need for JSON parsing in the updater. ScannerV4 will pull the zip bundle and populate the Matcher DB with the enrichment data.
```

Comment:
> The CVSS Updater in Central will be modified to download zip files corresponding to four distinct severity levels and write them into a single zip file stored in Central's file system.

In other words, you're making one  HTTP request for each `<severity>.zip`. What happens if one of these GETs fails when generating the single zip?

Reply:
We need to make multiple requests for each severity.

Reply:
> We need to make multiple requests for each severity.

Let's break it down.

Success scenario:

| What | Result | What to do |
|-|-|-|
| GET /bucket/<first>-severity.zip | Success | Keep in memory to add to the final zip file |
| GET /bucket/<second>-severity.zip | Success | Keep in memory to add to the final zip file |
| GET /bucket/<third>-severity.zip | Success |  Keep in memory to add to the final zip file |
| Create a final zip file from all zips in memory | Success | Write zip file to the filesystem and serve upon request from Scanner. |

Scenario I want to discuss:

| What | Result | What to do |
|-|-|-|
| GET /bucket/<first>-severity.zip | Success | Keep in memory to add to the final zip file |
| GET /bucket/<second>-severity.zip | Success | Keep in memory to add to the final zip file |
| GET /bucket/<third>-severity.zip | Failed (e.g. 401, 403, 505, Timeout) |  A |
| Create a final zip file from all zips in memory | B | C |

Could you please fill A, B and C?

Reply:
| Action                                                   | Result                                         | Next Step                               |
|----------------------------------------------------------|------------------------------------------------|-----------------------------------------|
| GET /bucket/1st-severity.json                            | Success                                        | download to a tmp directory (disk)      |
| GET /bucket/2nd-severity.zip                             | Success                                        | download to same tmp directory (disk)   |
| GET /bucket/3rd-severity.zip                             | Failed (e.g. 401, 403, 505, Timeout)           | SKIP                                    |
| Create a final zip file from all jsons in zip files      | Success                                        | Wait 4 hours for another update            |

Reply:
Thanks, worth adding to the doc.

Can you confirm if the enricher in ClairCore will continue to serve the previous well-known score for the severity that failed?

Also, worth checking if "download to a tmp/ directory (tmp)" in Central is fine. Just be sure.

Reply:
> Thanks, worth adding to the doc.
> 
> Can you confirm if the enricher in ClairCore will continue to serve the previous well-known score for the severity that failed?
> 
> Also, worth checking if "download to a tmp/ directory (tmp)" in Central is fine. Just be sure.

I am pretty sure we can create tmp directory in central, I've tried a few times in other PRs but it's allowed to create regular directory. But the tmp directory eventually need to go away after we generate the single zip file so it should be ok.

Reply:
> I am pretty sure we can create tmp directory in central, I've tried a few times in other PRs but it's allowed to create regular directory.

Thank you for checking :+1:

> so their availability should be consistently high.

This will run in customer environments, so it might not only depend on Google but on customer network setup, proxies, etc. Hence my concern on handling network issues gracefully.

@daynewlee:

> @jvdm:
>
> > Can you confirm if the enricher in ClairCore will continue to serve the previous well-known score for the severity that failed?
>
> I looked at the updateEnrichment() method in [Claircore](https://github.com/quay/claircore/blob/main/datastore/postgres/enrichment.go), seems like they intentionally block the old data after every update:

It makes sense for them to block what was updated. The scenario I am thinking of is we skip one severity (like in the example table, where we said "3rd" was skipped). But when updating the enrichments, ClairCore will mark the previous severities of the 3rd category as "outdated", effectively not showing them anymore in vulnerability matching results until we update them in the next enrichment cycle.

In simpler terms, do we need to take extra steps to avoid overwriting or blocking prior severity information when adding enrichments to ClairCore?

Reply:
Curious, how did you come up with this scheme (to query by severity)?